src/physics/base/ImportedProcessAdapter.cc

```diff
@@ -0,0 +1,205 @@
+    }
+
+    CELER_ENSURE(processes_.size() == ids_.size());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Return physics tables for a particle type and process.
+ *
+ * Returns 'invalid' ID if process is not present for the given particle type.
+ */
+auto ImportedProcesses::find(key_type particle_process) const -> ImportProcessId
+{
+    auto iter = ids_.find(particle_process);
+    if (iter == ids_.end())
+        return {};
+
+    return iter->second;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from shared process data.
+ */
+ImportedProcessAdapter::ImportedProcessAdapter(SPConstImported imported,
+                                               const SPConstParticles& particles,
+                                               ImportProcessClass process_class,
+                                               SpanConstPDG       pdg_numbers)
+    : imported_(std::move(imported))
+{
+    CELER_EXPECT(particles);
+    CELER_EXPECT(!pdg_numbers.empty());
+    for (PDGNumber pdg : pdg_numbers)
+    {
+        ParticleProcessIds proc_ids;
+        proc_ids.process = imported_->find({pdg, process_class});
+        CELER_VALIDATE(proc_ids.process,
+                       "Imported process data is unavalable for '"
+                           << to_cstring(process_class) << "' for PDG{"
+                           << pdg.get() << "}");
+
+        // Loop through available tables
+        const auto& tables = imported_->get(proc_ids.process).tables;
+        for (auto table_id : range(ImportTableId(tables.size())))
+        {
+            // Map table types to IDs in our imported data
+            ImportTableId* dst = nullptr;
+            switch (tables[table_id.get()].table_type)
+            {
+                case ImportTableType::dedx:
+                    dst = &proc_ids.dedx;
+                    break;
+                case ImportTableType::range:
+                    dst = &proc_ids.range;
+                    break;
+                case ImportTableType::lambda:
+                    dst = &proc_ids.lambda;
+                    break;
+                case ImportTableType::lambda_prim:
+                    dst = &proc_ids.lambda_prim;
+                    break;
+                default:
+                    //  Not a data type we care about
+                    continue;
+            }
+            CELER_ASSERT(dst);
+            CELER_VALIDATE(!*dst,
+                           "Duplicate table type '"
+                               << to_cstring(tables[table_id.get()].table_type)
+                               << "' in process data for '"
+                               << to_cstring(process_class) << "'");
+            *dst = table_id;
+        }
+
+        auto particle_id = particles->find(pdg);
+        CELER_VALIDATE(particle_id,
+                       "Parrticle PDG{" << pdg.get() << "} is not available");
```

Comment:
```suggestion
                       "Particle PDG{" << pdg.get() << "} is not available");
```

Reply:
My poor macbook's keys have been super sticky lately :(

Reply:
You know what?  I had the very same 'r' key sticking in my laptop.  There is a known problem with mac keyboards, and I had mine replaced by apple.  Of course, I had Fermilab send it to Apple, and it took a bit more than a week to get it back.

Reply:
https://support.apple.com/keyboard-service-program-for-mac-notebooks

Reply:
My delete key has been doubling too, so now I'm getting stuck in lots of infinite loops :(